### ISSUE TITLE
Proj0033 false positive: $(MSBuildThisFileDirectory) (and other MSBuild properties) not expanded before file-existence check

### ISSUE THREAD

**Issue by JohnVerheij** [dotnet-project-file-analyzers/dotnet-project-file-analyzers#778]
`Lib/Lib.csproj`:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="$(MSBuildThisFileDirectory)..\My.Analyzer\My.Analyzer.csproj"
                      PrivateAssets="all"
                      ReferenceOutputAssembly="false"
                      OutputItemType="Analyzer" />
  </ItemGroup>
</Project>
```

`My.Analyzer/My.Analyzer.csproj` exists and the `Include` resolves correctly at build time
(MSBuild expands `$(MSBuildThisFileDirectory)` to the absolute path of the directory holding the
`.csproj`). The project builds, links, and runs. Yet:

```
Lib\Lib.csproj(6,5): warning Proj0033: The Include '$(MSBuildThisFileDirectory)..\My.Analyzer\My.Analyzer.csproj' of <ProjectReference> does not exist
```

The error message quotes the unexpanded property literally, suggesting the existence check runs
against the raw XML attribute value rather than the MSBuild-evaluated path. References that use
plain relative paths (`..\My.Analyzer\My.Analyzer.csproj`) do not fire the rule.

The same blind-spot also makes the rule fire on `<ProjectReference>` items inside an
`<ItemGroup Condition="…">` even when the Condition would exclude the item — likely the same
"don't evaluate, just check the literal text" code path.

Closest open precedent: #756 (Proj0044 with a Condition) and the closed-fixed sibling fixes in
#258 (Proj0009) and #257 (Proj0028).

Workaround: `<NoWarn>$(NoWarn);Proj0033</NoWarn>`.

**Comment by Corniel** [comment]
We are ware of this, and have an (old) issue for this: #33.

That being said, `$(MSBuildThisFileDirectory)` should be doable to take into consideration, so there are plans to pick this up quickly.

**Comment by JohnVerheij** [comment]
@Corniel, if a contribution from outside would be welcome on this one, I'd be happy to help. To give you a feel for the direction I'd take, I worked through the options first:

- **Narrow path**: small helper that resolves the reserved per-file MSBuild properties (`$(MSBuildThisFileDirectory)`, `$(MSBuildProjectDirectory)`) before the existence check. Statically derivable from the file's location, no consumer opt-in needed. Applies cleanly to `ProjectReferenceChecker` (which would also unstick `Proj0051` on the same lookup) and to the parallel pattern in `BuildActionIncludeShouldExist`.
- **`<CompilerVisibleProperty>`** (your #33 note): right mechanism for surfacing *user-defined* properties to analyzers later, but it doesn't help here. `$(MSBuildThisFileDirectory)` isn't a property the consumer would expose; MSBuild derives it per-file.
- **Embedded MSBuild evaluation / `Microsoft.Build` dependency**: more accurate but too heavy for the scope.

If you'd rather handle it yourself or want a different direction, no problem. Just say the word.

**Comment by Corniel** [comment]
The biggest problem with `<CompilerVisibleProperty>` currently is that we can not test it automatically, as Buildalyzer does not support that yet (see https://github.com/Buildalyzer/Buildalyzer/pull/301).

So currently, all `<CompilerVisibleProperty>` reads have a fallback (which is a good practice anyway).

The best place to try to resolve `IOFFile`'s based on these settings is probably : [DotNetProjectFile.Analyzers/Extensions/DotNetProjectFile.Diagnostics.ProjectFileAnalysisContext.cs](https://github.com/dotnet-project-file-analyzers/dotnet-project-file-analyzers/blob/main/src/DotNetProjectFile.Analyzers/Extensions/DotNetProjectFile.Diagnostics.ProjectFileAnalysisContext.cs).

We're definitely open to PR's.

**Comment by JohnVerheij** [comment]
Thanks for the pointer to `ProjectFileAnalysisContext.cs` and for the Buildalyzer context. I'll open a PR with a `Resolve` / `Files` extension pair that handles `$(MSBuildThisFileDirectory)` and `$(MSBuildProjectDirectory)`, applied to `ProjectReferenceChecker` (which also unsticks Proj0051 for these inputs) and `BuildActionIncludeShouldExist`. Includes containing other unresolved properties fall through to the existing literal lookup, keeping the broader `<CompilerVisibleProperty>` direction separate as you noted.
